src/task_manager.rs

```diff
@@ -0,0 +1,335 @@
+                Some(ref c) => c(TypeResponse::DefId(task_id, pos.line as usize * 10)),
+                None => {}
+            },
+            TypeMessage::TypeForDefId(task_id, def_id) => match self.send_channel {
+                Some(ref c) => c(TypeResponse::Type(task_id, format!("<type:{}>", def_id))),
+                None => {}
+            },
+            TypeMessage::CompletionsForDefId(task_id, def_id) => match self.send_channel {
+                Some(ref c) => c(TypeResponse::Completions(
+                    task_id,
+                    vec![
+                        ("bar".into(), format!("First option for {}", def_id)),
+                        ("foo".into(), format!("Second option for {}", def_id)),
+                    ],
+                )),
+                None => {}
+            },
+        }
+    }
+}
+
+impl FakeTypeChecker {
+    fn new() -> FakeTypeChecker {
+        FakeTypeChecker { send_channel: None }
+    }
+}
+
+pub struct TaskManager {
+    live_recipes: HashMap<TaskId, Vec<RecipeStep>>,
+    receive_channel: Receiver<MsgToManager>,
+
+    //Channel to send to this task manager
+    pub send_to_manager: Sender<MsgToManager>,
+
+    //Channels to communicate with other subsystems
+    send_to_type_checker: Option<Sender<MsgFromManager<TypeMessage>>>,
+    send_to_lsp_responder: Option<Sender<MsgFromManager<LspResponse>>>,
+
+    //Handles when we shutdown threads,
+    handle_for_type_checker: Option<std::thread::JoinHandle<()>>,
+    handle_for_lsp_responder: Option<std::thread::JoinHandle<()>>,
+}
+
+impl TaskManager {
+    pub fn new() -> TaskManager {
+        let (host_tx, host_rx) = channel();
+
+        TaskManager {
+            live_recipes: HashMap::new(),
+            receive_channel: host_rx,
+            send_to_manager: host_tx,
+            send_to_type_checker: None,
+            send_to_lsp_responder: None,
+            handle_for_type_checker: None,
+            handle_for_lsp_responder: None,
+        }
+    }
+
+    fn send_next_step(&mut self, task_id: TaskId, argument: Box<dyn std::any::Any>) {
+        match self.live_recipes.get_mut(&task_id) {
+            Some(x) => {
+                if x.len() > 0 {
+                    let next_step = x.remove(0);
+
+                    match next_step {
+                        RecipeStep::GetDefIdForPosition => {
+                            if let Ok(position) = argument.downcast::<Position>() {
+                                self.send_to_type_checker.as_ref().map(|x| {
+                                    x.send(MsgFromManager::Message(TypeMessage::DefIdForPos(
+                                        task_id, *position,
+                                    )))
+                                    .unwrap()
+                                });
+                            } else {
+                                unimplemented!("Internal error: malformed GetDefIdForPosition");
```

Comment:
Nit: I think the intent of `unimplemented` is to be used for things that are not implemented *yet* -- I would expect this to just be a `panic!`